services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/RequestBatchBuffer.java

```diff
@@ -0,0 +1,137 @@
+     * to idToBatchContext had an id of 22, nextBatchEntry will have a value of 23).
+     */
+    private int nextBatchEntry;
+
+    /**
+     * The scheduled flush tasks associated with this batchBuffer.
+     */
+    private ScheduledFuture<?> scheduledFlush;
+
+    public RequestBatchBuffer(int maxBufferSize, ScheduledFuture<?> scheduledFlush) {
+        this.idToBatchContext = new ConcurrentHashMap<>();
+        this.maxBufferSize = maxBufferSize;
+        this.nextId = 0;
+        this.nextBatchEntry = 0;
+        this.scheduledFlush = scheduledFlush;
+    }
+
+    public Map<String, BatchingExecutionContext<RequestT, ResponseT>> flushableRequests(int maxBatchItems) {
+        synchronized (flushLock) {
+            if (idToBatchContext.size() >= maxBatchItems) {
+                return extractFlushedEntries(maxBatchItems);
+            }
+            return new ConcurrentHashMap<>();
+        }
+    }
+
+    public Map<String, BatchingExecutionContext<RequestT, ResponseT>> flushableScheduledRequests(int maxBatchItems) {
+        synchronized (flushLock) {
+            if (idToBatchContext.size() > 0) {
+                return extractFlushedEntries(maxBatchItems);
+            }
+            return new ConcurrentHashMap<>();
+        }
+    }
+
+    private Map<String, BatchingExecutionContext<RequestT, ResponseT>> extractFlushedEntries(int maxBatchItems) {
+        LinkedHashMap<String, BatchingExecutionContext<RequestT, ResponseT>> requestEntries = new LinkedHashMap<>();
+        String nextEntry;
+        while (requestEntries.size() < maxBatchItems && hasNextBatchEntry()) {
+            nextEntry = nextBatchEntry();
+            requestEntries.put(nextEntry, idToBatchContext.get(nextEntry));
+            idToBatchContext.remove(nextEntry);
+        }
+        return requestEntries;
+    }
+
+    public void put(RequestT request, CompletableFuture<ResponseT> response) {
+        synchronized (this) {
+            if (idToBatchContext.size() == maxBufferSize) {
+                throw new IllegalStateException("Reached MaxBufferSize of: " + maxBufferSize);
+            }
+
+            if (nextId == Integer.MAX_VALUE) {
+                nextId = 0;
+            }
+            String id = Integer.toString(nextId++);
+            idToBatchContext.put(id, new BatchingExecutionContext<>(request, response));
+        }
+    }
+
+    private boolean hasNextBatchEntry() {
+        return idToBatchContext.containsKey(Integer.toString(nextBatchEntry));
+    }
+
+    private String nextBatchEntry() {
+        if (nextBatchEntry == Integer.MAX_VALUE) {
+            nextBatchEntry = 0;
+        }
+        return Integer.toString(nextBatchEntry++);
+    }
+
+    public void putScheduledFlush(ScheduledFuture<?> scheduledFlush) {
+        this.scheduledFlush = scheduledFlush;
+    }
+
+    public void cancelScheduledFlush() {
+        scheduledFlush.cancel(false);
```

Comment:
This class still seems a little weird to me since it tells you what entries are flushable, and knows about removes about entries that are about to be flushed, but doesn't actually perform the flushing.

Probably not easy to refactor this without prolonging this review though. Can we add a tracking item to possibly refactor this in the future?

Reply:
Created `JAVA-7832` for this. Current IM was based on reusing as much as possible from the Project branch , we will need more time to refactor this , created a task for this , WDYT if we refactor this class after the release , else it will delay the release of the feature